tardis/energy_input/util.py

```diff
@@ -0,0 +1,261 @@
+import astropy.units as u
+import tardis.constants as const
+import numpy as np
+from astropy.coordinates import spherical_to_cartesian
+
+R_ELECTRON = const.a0.cgs * const.alpha.cgs ** 2.0
+KEV2ERG = (1000 * u.eV).to("erg").value
+
+
+class SphericalVector(object):
+    """
+    Direction object to hold spherical polar and Cartesian directions
+    Must be initialized with r, theta, phi
+
+    Attributes
+    ----------
+    r : float64
+             vector r position
+    theta : float64
+             vector mu position
+    phi : float64
+             vector phi position
+    mu : float64
+             calculated vector theta position
+    x : float64
+             calculated vector x position
+    y : float64
+             calculated vector y position
+    z : float64
+             calculated vector z position
+    """
+
+    def __init__(self, r, theta, phi=0.0):
+        self.r = r
+        self.theta = theta
+        self.phi = phi
+
+    @property
+    def mu(self):
+        return np.cos(self.theta)
+
+    @property
+    def get_cartesian_coords(self):
+        # 0.5*np.pi subtracted because of the definition of theta
+        # in astropy.coordinates.cartesian_to_spherical
+        x, y, z = spherical_to_cartesian(
+            self.r, self.theta - 0.5 * np.pi, self.phi
+        )
+        return x.value, y.value, z.value
+
+
+def kappa_calculation(energy):
+    """
+    Calculates kappa for various other calculations
+    i.e. energy normalized to electron rest energy
+    511.0 KeV
+
+    Parameters
+    ----------
+    energy : dtype float
+
+    Returns
+    -------
+    kappa : dtype float
+
+    """
+    k = energy / 511.0
+    return k
+
+
+def euler_rodrigues(theta, direction):
```

Comment:
This seems like it could potentially be imported from other python packages. I have not looked at it in detail but scipy.spatial.transform.Rotation looks promising.